src/mlpack/methods/preprocess/preprocess_describe_main.cpp

```diff
@@ -213,21 +232,39 @@ static void mlpackMain()
         % Kurtosis(feature, fStd, fMean, population)
         % StandardError(feature.n_elem, fStd)
         << endl;
+
+    // Store the output in the output arma::mat.
+    outputMat.row(dim) = arma::vec({
```

Comment:
@PlantsAndBuildings Would you mind, justifying the rationale behind the need for output matrix for data description? I think there is no necessity of having output matrix, in this case, just the print should do the required. Moreover, in the current state, the user will not able to comprehend what does each output cell signifies (As you will be getting a column of 11 floating point values in the output file), if he/she went looking for the results in the output file.

Reply:
Thanks for the review.

I'll start off by saying that - yes, you're absolutely right that the user will not be able to make sense of the output. Also, I know that what I'm about to describe is not a very clean way to do things - and I'm open to suggestions to do things differently. 

I added that param because I don't see any other way to access the results  programatically - which is a problem if you're trying to write tests for this binding. I'll elaborate - consider the following test which checks the working of --row_major command line flag:
``` c++
 BOOST_AUTO_TEST_CASE(PreprocessDescribeRowMajorTest)
 {
   arma::mat dataset;
   data::Load("trainSet.csv", dataset);

   SetInputParam("input", std::move(dataset));
   mlpackMain();

   arma::mat colMajorOutput = CLI::GetParam<arma::mat>("output");

   SetInputParam("row_major", true);
   mlpackMain();

   arma::mat rowMajorOutput = CLI::GetParam<arma::mat>("output");

   BOOST_REQUIRE_EQUAL(colMajorOutput.n_rows,
                       CLI::GetParam<arma::mat>("input").n_rows);
   BOOST_REQUIRE_EQUAL(rowMajorOutput.n_rows,
                       CLI::GetParam<arma::mat>("input").n_cols);
   BOOST_REQUIRE_EQUAL(colMajorOutput.n_cols, rowMajorOutput.n_cols);
 }
```
I'm not sure how we would able to get `colMajorOutput` and `rowMajorOutput` without having to add the `output` param. Let me know what you think.

Reply:
This is definitely a hard binding to test.  I have to agree with Manish, it looks like the user will not be able to understand the output of the output matrix.  I think this will be a bit confusing.  I do have another idea, let me know what you think:

The output all goes to `Log::Info`.  But we can hijack the destination; this is done in the tests for `PrefixedOutStream`:

```
std::ostringstream oss;
Log::Info.destination = oss;

// Call mlpackMain() and the output will get filled up...
mlpackMain();

// Store the output.
const std::string output = oss.str();
oss.clear();

// Now change some parameters and run again...
...
mlpackMain();

// Now we can check that the output was different.
BOOST_REQUIRE_NE(oss.str(), output);
```

What I showed above only shows how to check if the output of two runs is different.  I think it would be possible and not too hard to actually parse the output, but honestly in most cases here I think just checking that the output is different is sufficient (like for the `row_major` test you have above).

Reply:
```
std::ostringstream oss;
Log::Info.destination = oss;
```
Unfortunately, this is not allowed. Refer [this](http://en.cppreference.com/w/cpp/io/basic_ostream/operator%3D). I'm looking into this a bit more - but I don't see too many alternatives other than what Manish has suggested (the boundary checks approach).

Reply:
You are right, sorry, `destination` is a reference.  Do you think the following might work?

```
PrefixedOutStream oldInfo = Log::Info;
std::ostringstream oss;
Log::Info = PrefixedOutStream(oss, ""); // no prefix needed for this.

// Call mlpackMain(), do other stuff...
mlpackMain();
...

// Now reset what Log::Info was.
Log::Info = oldInfo;
```

I think that will work but I am not sure...

Reply:
Sorry for the slow response on this. No, this too will not work. It says:
```
/root/mlpack/src/mlpack/core/util/prefixedoutstream.hpp:46:7: error: non-static reference member 'std::ostream& mlpack::util::PrefixedOutStream::destination', can't use default assignment operator
```

Reply:
You are right, I did not think that through.  I can supply a patch that will internally change `destination` to `std::ostream*`; that could help.  Would you like me to do that?

Reply:
Umm .. okay. I do have a few concerns though.

It just seems like a lot of effort (on your end, since you're the one who will have to figure out how to write the patch) to go through for testing just one binding. Will doing this (ie. changing `destination` to `std::ostream*`) be useful for some other stuff as well? Rather, are there any more use-cases where we would need to hijack the INFO stream? If not, then I guess we should consider going with one of the other alternatives:

1. What Manish suggested initially - the boundary checks approach.
2. Writing explicitly to an output file from `preprocess_describe_main.cpp` binding - this way we can add column headings to the output file and the output would make sense to the user (see previous comments).
3. "Hiding" the output parameter -- something like `PARAM_MAT_OUT("__output")` and explicitly warning the user that this option is meant for internal use only.
4. This might be a bit drastic but - scrapping the tests that need to look at the output of the binding? - since the binding is obviously working fine.

Either way, if you think that providing the patch (or perhaps holding my hand through writing it) is relatively easy and worth the effort - then that is what we should go with. Let me know, I'm happy to go with any approach.

Reply:
Sorry for the slow response on this one, it has been a busy past few days and I have fallen far behind.

Honestly I don't have a huge problem with option 4.  I agree that it is giving us output as expected.  Manish's suggestion to check the number of dimensions sounds just fine to me.

I do think the binding is more useful when it provides output on stdout instead of into a file.  If we want to change the output, I might think a better way to do it would be to use `PARAM_STRING_OUT()` or `PARAM_VECTOR_OUT()` or something like this to provide the output.  But we could open a separate issue for that if there was interest in solving it.  I don't think it's a hugely pressing need so I don't think we need to do that here.

So, up to you---if you'd rather take suggestions (4) and (1), that would be just fine with me.  But if you want to do something more complicated, we can do that also.